suite/spdz2k/src/main/java/dk/alexandra/fresco/suite/marlin/protocols/computations/BroadcastComputation.java

```diff
@@ -0,0 +1,52 @@
+package dk.alexandra.fresco.suite.marlin.protocols.computations;
+
+import dk.alexandra.fresco.framework.DRes;
+import dk.alexandra.fresco.framework.builder.Computation;
+import dk.alexandra.fresco.framework.builder.ProtocolBuilderImpl;
+import dk.alexandra.fresco.suite.marlin.protocols.natives.AllBroadcastProtocol;
+import dk.alexandra.fresco.suite.marlin.protocols.natives.BroadcastValidationProtocol;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Generic active broadcast validation computation. <p>Uses generic native protocols {@link
+ * AllBroadcastProtocol} and {@link BroadcastValidationProtocol}.</p>
+ */
+public class BroadcastComputation<BuilderT extends ProtocolBuilderImpl<BuilderT>> implements
+    Computation<List<byte[]>, BuilderT> {
+
+  private static final Logger logger = LoggerFactory.getLogger(Spdz2kCommitmentComputation.class);
+  private final List<byte[]> input;
+
+  BroadcastComputation(List<byte[]> input) {
+    this.input = input;
+  }
+
+  BroadcastComputation(byte[] input) {
+    this(java.util.Collections.singletonList(input));
+  }
+
+  @Override
+  public DRes<List<byte[]>> buildComputation(BuilderT builder) {
+    return builder.par(par -> {
+      List<DRes<List<byte[]>>> broadcastValues = new LinkedList<>();
+      for (byte[] singleInput : input) {
+        broadcastValues.add(par.append(new AllBroadcastProtocol<>(singleInput)));
+      }
+      return () -> broadcastValues;
+    }).seq((seq, lst) -> {
+      List<byte[]> toValidate = lst.stream()
+          .flatMap(broadcast -> broadcast.out().stream())
+          .collect(Collectors.toList());
+      DRes<Void> nothing = seq.append(new BroadcastValidationProtocol<>(toValidate));
+      return () -> {
+        nothing.out();
```

Comment:
this is a very code construct. It seems it is written this way to make testing simpler.

There are better ways to do that.

Reply:
The problem is that the BroadcastValidationProtocol does not return anything, however, in order for it be invoked I need to call .out in a subsequent scope. I can have it return a Boolean instead of throwing a malicious exception and check the returned value in the next scope. Alternatively, I can move the () -> nothing.out() construct into the BroadcastValidationProtocol class but then it would need to become a computation instead of a native protocol.